ev3dev/core.py

```diff
@@ -2615,27 +2610,121 @@ def __init__(self, address=None, name_pattern=SYSTEM_DEVICE_NAME_CONVENTION, nam
     #: Calibration ???
     MODE_IR_CAL = 'IR-CAL'
 
-
     MODES = (
-      'IR-PROX',
-      'IR-SEEK',
-      'IR-REMOTE',
-      'IR-REM-A',
-      'IR-CAL',
+      MODE_IR_PROX,
+      MODE_IR_SEEK,
+      MODE_IR_REMOTE,
+      MODE_IR_REM_A,
+      MODE_IR_CAL
     )
 
+    _BUTTON_VALUES = {
+            0: [],
+            1: ['red_up'],
+            2: ['red_down'],
+            3: ['blue_up'],
+            4: ['blue_down'],
+            5: ['red_up', 'blue_up'],
+            6: ['red_up', 'blue_down'],
+            7: ['red_down', 'blue_up'],
+            8: ['red_down', 'blue_down'],
+            9: ['beacon'],
+            10: ['red_up', 'red_down'],
+            11: ['blue_up', 'blue_down']
+            }
+
+    #: Handles ``Red Up`` events.
+    on_red_up = None
+
+    #: Handles ``Red Down`` events.
+    on_red_down = None
+
+    #: Handles ``Blue Up`` events.
+    on_blue_up = None
+
+    #: Handles ``Blue Down`` events.
+    on_blue_down = None
+
+    #: Handles ``Beacon`` events.
+    on_beacon = None
+
+    def __init__(self, address=None, name_pattern=SYSTEM_DEVICE_NAME_CONVENTION, name_exact=False, **kwargs):
+        super(InfraredSensor, self).__init__(address, name_pattern, name_exact, driver_name=['lego-ev3-ir'], **kwargs)
 
     @property
     def proximity(self):
         """
         A measurement of the distance between the sensor and the remote,
         as a percentage. 100% is approximately 70cm/27in.
         """
+        self.mode = self.MODE_IR_PROX
+        return self.value(0)
 
-        if self.auto_mode:
-            self.mode = self.MODE_IR_PROX
+    # BeaconSeeker
+    def heading(self, channel=1):
+        """
+        Returns heading (-25, 25) to the beacon on the given channel.
+        """
+        self.mode = self.MODE_IR_SEEK
+        channel = max(1, min(4, channel)) - 1
+        return self.value(channel * 2)
+
+    def distance(self, channel=1):
+        """
+        Returns distance (0, 100) to the beacon on the given channel.
+        Returns -128 when beacon is not found.
```

Comment:
Well duh, _of course_ it returns -128 :laughing: Do we want to catch that and return `None` instead? I think no one actually reads our docs :wink:

Reply:
Not sure...what does the EV3-G block do in this scenario?

Reply:
Returning 'None' makes sense.

Reply:
ack